### PR TITLE
Ensure that swap with self is safe

### PR DESCRIPTION
**Pull request by jbcoe**
Tests to make sure that self-swap and self-assign should do not make an object valueless.

### CODE REVIEW

**Comment by jbcoe** [comment]
Getting formatting right from the phone is hard.

File: polymorphic.h [hunk 1]
```diff
@@ -272,6 +272,10 @@ class polymorphic {
       std::allocator_traits<A>::is_always_equal::value) {
     assert(other.cb_ != nullptr);  // LCOV_EXCL_LINE
 
+    if (this == &other) {
+      return;
+    }
+
     if constexpr (allocator_traits::propagate_on_container_swap::value) {
       // If allocators move with their allocated objects we can swap both.
```
Comment (nbx8) [inline polymorphic.h:280]: // If allocators move with their allocated objects, m we can swap both.

File: DRAFT.md [hunk 2]
```diff
@@ -672,7 +672,7 @@ constexpr indirect& operator=(const indirect& other);
 
 * _Preconditions_: `other` is not valueless.
 
-* _Effects_: If
+* _Effects_: If `this==&other`, then does nothing. Otherwise if
```
Comment (nbx8) [inline DRAFT.md:675]: Otherwise, if

File: DRAFT.md [hunk 3]
```diff
@@ -749,7 +749,8 @@ constexpr void swap(indirect& other) noexcept(
 
 * _Preconditions_: `*this` is not valueless, `other` is not valueless.
 
-* _Effects_: Swaps the objects owned by `*this` and `other`. If
+* _Effects_: If `this==&other`, then does nothing. Otherwise swaps the objects
```
Comment (nbx8) [inline DRAFT.md:752]: Otherwise, swaps

File: DRAFT.md [hunk 4]
```diff
@@ -1166,8 +1167,9 @@ constexpr polymorphic& operator=(const polymorphic& other);
 
 * _Preconditions_: `other` is not valueless.
 
-* _Effects_: If `*this` is not valueless, destroys the owned object. If
-`allocator_traits<allocator_type>::propagate_on_container_copy_assignment::value
+* _Effects_: If `this==&other`, then does nothing. Otherwise if `*this` is not
```
Comment (nbx8) [inline DRAFT.md:1170]: Otherwise, if

File: DRAFT.md [hunk 5]
```diff
@@ -1237,7 +1239,8 @@ constexpr void swap(polymorphic& other) noexcept(
 
 * _Preconditions_: `*this` is not valueless, `other` is not valueless.
 
-* _Effects_: Swaps the objects owned by `*this` and `other`. If
+* _Effects_: If `this==&other`, then does nothing. Otherwise swaps the objects
```
Comment (nbx8) [inline DRAFT.md:1242]: Otherwise, swaps

**Review by nbx8 — APPROVED** [review]

**Review by Twon — APPROVED** [review]

**Review by Twon — APPROVED** [review]

**Review by Twon — APPROVED** [review]

**Pull request merged**